File: frontend/src/types/roadmap.ts
export type Difficulty = 'easy' | 'medium' | 'hard' | 'very_hard'
export type PriorityLevel = 'high' | 'medium' | 'low'
export type NodeLevel = 'topic' | 'subtopic' | 'concept'
export type ProgressStatus = 'not_started' | 'in_progress' | 'completed' | 'needs_revision'

export interface TopicProgress {
  status: ProgressStatus
  completion_percent: number
  retention_percent: number | null
  accuracy_percent: number | null
  pyqs_solved: number
  practice_solved: number
  revision_count: number
  time_spent_minutes: number
  last_studied_at: string | null
  last_revised_at: string | null
  completed_at: string | null
}

export interface TopicNode {
  id: number
  slug: string
  name: string
  level: NodeLevel
  order_index: number
  difficulty: Difficulty | null
  importance_1to5: number | null
  pyq_frequency: string | null
  estimated_hours: number | null
  revision_frequency: string | null
  prerequisites_text: string | null
  progress: TopicProgress
  children: TopicNode[]
}

export interface SubjectProgressSummary {
  total: number
  completed: number
  in_progress: number
  not_started: number
  needs_revision: number
  avg_completion_percent: number
}

export interface SubjectNode {
  id: number
  slug: string
  name: string
  description: string | null
  weightage_min_percent: number | null
  weightage_max_percent: number | null
  difficulty: Difficulty | null
  priority: PriorityLevel | null
  order_index: number
  is_official_section: boolean
  progress_summary: SubjectProgressSummary
  topics: TopicNode[]
}

export interface ProgressUpdateRequest {
  status?: ProgressStatus
  completion_percent?: number
  time_spent_minutes_delta?: number
}

File: backend/app/schemas/roadmap.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.enums import Difficulty, NodeLevel, PriorityLevel, ProgressStatus


class TopicProgressRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    status: ProgressStatus
    completion_percent: float
    retention_percent: float | None
    accuracy_percent: float | None
    pyqs_solved: int
    practice_solved: int
    revision_count: int
    time_spent_minutes: int
    last_studied_at: datetime | None
    last_revised_at: datetime | None
    completed_at: datetime | None


DEFAULT_PROGRESS = TopicProgressRead(
    status=ProgressStatus.NOT_STARTED,
    completion_percent=0.0,
    retention_percent=None,
    accuracy_percent=None,
    pyqs_solved=0,
    practice_solved=0,
    revision_count=0,
    time_spent_minutes=0,
    last_studied_at=None,
    last_revised_at=None,
    completed_at=None,
)


class TopicNode(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    slug: str
    name: str
    level: NodeLevel
    order_index: int
    difficulty: Difficulty | None
    importance_1to5: int | None
    pyq_frequency: str | None
    estimated_hours: float | None
    revision_frequency: str | None
    prerequisites_text: str | None
    progress: TopicProgressRead
    children: list["TopicNode"] = []


class SubjectProgressSummary(BaseModel):
    total: int
    completed: int
    in_progress: int
    not_started: int
    needs_revision: int
    avg_completion_percent: float


class SubjectNode(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    slug: str
    name: str
    description: str | None
    weightage_min_percent: float | None
    weightage_max_percent: float | None
    difficulty: Difficulty | None
    priority: PriorityLevel | None
    order_index: int
    is_official_section: bool
    progress_summary: SubjectProgressSummary
    topics: list[TopicNode]


class ProgressUpdateRequest(BaseModel):
    status: ProgressStatus | None = None
    completion_percent: float | None = None
    time_spent_minutes_delta: int | None = None
